// stl-parser/build/Ascii.js
// Generated by CoffeeScript 1.10.0
(function() {
  var Ascii;

  Ascii = (function() {
    var skipWhitespaces, whitespaces;

    whitespaces = [' ', '\r', '\n', '\t', '\v', '\f'];

    skipWhitespaces = function() {
      var results, skip;
      skip = true;
      results = [];
      while (skip) {
        if (this.currentCharIsWhitespace() && !this.reachedEnd()) {
          results.push(this.index++);
        } else {
          results.push(skip = false);
        }
      }
      return results;
    };

    function Ascii(fileContent) {
      this.content = fileContent;
      this.index = 0;
    }

    Ascii.prototype.nextText = function() {
      skipWhitespaces.call(this);
      return this.readUntilWhitespace();
    };

    Ascii.prototype.currentChar = function() {
      return this.content[this.index];
    };

    Ascii.prototype.currentCharIsWhitespace = function() {
      var i, len, space;
      for (i = 0, len = whitespaces.length; i < len; i++) {
        space = whitespaces[i];
        if (this.currentChar() === space) {
          return true;
        }
      }
      return false;
    };

    Ascii.prototype.readUntilWhitespace = function() {
      var readContent;
      readContent = '';
      while (!this.currentCharIsWhitespace() && !this.reachedEnd()) {
        readContent = readContent + this.currentChar();
        this.index++;
      }
      return readContent;
    };

    Ascii.prototype.reachedEnd = function() {
      return this.index === this.content.length;
    };

    return Ascii;

  })();

  module.exports = Ascii;

}).call(this);
